File: go/cmd/api/accountDAO.go
package main

import (
	"database/sql"

	_ "github.com/lib/pq"
)

// Account represents the account data structure
type Account struct {
	AccountID string `json:"account_id"`
	Name      string `json:"name"`
	Email     string `json:"email"`
	Document  string `json:"document"`
	Password  string `json:"password"`
}

// IAccountDAO defines the interface for account data access operations
type IAccountDAO interface {
	Save(account *Account) error
	GetByID(accountID string) (*Account, error)
	GetByEmail(email string) (*Account, error)
}

// AccountDAODatabase implements IAccountDAO using PostgreSQL database
type AccountDAODatabase struct{}

func NewAccountDAODatabase() *AccountDAODatabase {
	return &AccountDAODatabase{}
}

func (dao *AccountDAODatabase) getDB() (*sql.DB, error) {
	connStr := "host=db port=5432 user=postgres password=postgres dbname=app sslmode=disable"
	return sql.Open("postgres", connStr)
}

func (dao *AccountDAODatabase) Save(account *Account) error {
	db, err := dao.getDB()
	if err != nil {
		return err
	}
	defer db.Close()

	query := "INSERT INTO ccca.account (account_id, name, email, document, password) VALUES ($1, $2, $3, $4, $5)"
	_, err = db.Exec(query, account.AccountID, account.Name, account.Email, account.Document, account.Password)
	return err
}

func (dao *AccountDAODatabase) GetByID(accountID string) (*Account, error) {
	db, err := dao.getDB()
	if err != nil {
		return nil, err
	}
	defer db.Close()

	query := "SELECT account_id, name, email, document, password FROM ccca.account WHERE account_id = $1"
	row := db.QueryRow(query, accountID)

	account := &Account{}
	err = row.Scan(&account.AccountID, &account.Name, &account.Email, &account.Document, &account.Password)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}

	return account, nil
}

func (dao *AccountDAODatabase) GetByEmail(email string) (*Account, error) {
	db, err := dao.getDB()
	if err != nil {
		return nil, err
	}
	defer db.Close()

	query := "SELECT account_id, name, email, document, password FROM ccca.account WHERE email = $1"
	row := db.QueryRow(query, email)

	account := &Account{}
	err = row.Scan(&account.AccountID, &account.Name, &account.Email, &account.Document, &account.Password)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}

	return account, nil
}

// AccountDAOMemory implements IAccountDAO using in-memory storage
type AccountDAOMemory struct {
	accounts   map[string]*Account
	emailIndex map[string]string
}

func NewAccountDAOMemory() *AccountDAOMemory {
	return &AccountDAOMemory{
		accounts:   make(map[string]*Account),
		emailIndex: make(map[string]string),
	}
}

func (dao *AccountDAOMemory) Save(account *Account) error {
	dao.accounts[account.AccountID] = account
	dao.emailIndex[account.Email] = account.AccountID
	return nil
}

func (dao *AccountDAOMemory) GetByID(accountID string) (*Account, error) {
	account, exists := dao.accounts[accountID]
	if !exists {
		return nil, nil
	}
	return account, nil
}

func (dao *AccountDAOMemory) GetByEmail(email string) (*Account, error) {
	accountID, exists := dao.emailIndex[email]
	if !exists {
		return nil, nil
	}

	account, exists := dao.accounts[accountID]
	if !exists {
		return nil, nil
	}

	return account, nil
}

File: node/src/DAO/accountDAO.ts
import pgPromise from "pg-promise";


export default interface IAccountDAO {
    save (account: any): Promise<void>;
    getById (accountId: string): Promise<any>;
    getByEmail(email: string): Promise<any>;
}

export class AccountDAODatabase implements IAccountDAO {
    async save(account: any): Promise<void> {
        const db = pgPromise()({
            host: "db",
            port: 5432,
            database: "app",
            user: "postgres",
            password: "postgres"
        });
        await db.query("insert into ccca.account (account_id, name, email, document, password) values ($1, $2, $3, $4, $5)", [account.accountId, account.name, account.email, account.document, account.password]);
        await db.$pool.end();
    }
    
    async getById(accountId: string): Promise<any> {
        const db = pgPromise()({
            host: "db",
            port: 5432,
            database: "app",
            user: "postgres",
            password: "postgres"
        });
        const account = await db.oneOrNone("select * from ccca.account where account_id = $1", [accountId]);
        await db.$pool.end();
        return account
    }

    async getByEmail(email: string): Promise<any> {
        const db = pgPromise()({
            host: "db",
            port: 5432,
            database: "app",
            user: "postgres",
            password: "postgres"
        });
        const account = await db.oneOrNone("select * from ccca.account where email = $1", [email]);
        await db.$pool.end();
        return account;
    }
}

export class AccountDAOMemory implements IAccountDAO {
    private accounts: Map<string, any> = new Map();
    private emailIndex: Map<string, string> = new Map();

    async save(account: any): Promise<void> {
        this.accounts.set(account.accountId, account);
        this.emailIndex.set(account.email, account.accountId);
    }

    async getById(accountId: string): Promise<any> {
        return this.accounts.get(accountId);
    }

    async getByEmail(email: string): Promise<any> {
        const accountId = this.emailIndex.get(email);
        if (accountId) {
            return this.accounts.get(accountId);
        }
        return null;
    }

}